### PR TITLE
Simple Prover App (depends on #32)

### PR DESCRIPTION
**Pull request by dtebbs**
Simple distributed prover app.  Reads binary proving key and assignment, and generates a grothBGM17 proof by calling the Distributed prover.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -35,6 +35,20 @@ jobs:
               run: |
                   scripts/local-cluster-setup.sh
                   scripts/dev-setup.sh
+            - name: Start local cluster
+              run: |
+                  docker-compose -f scripts/local_cluster/docker-compose.yml up -d
```
Comment (AntoineRondelet) [inline .github/workflows/main.yml:40]: Quick note here: `docker-compose.yml` is not a script. This is a configuration file, so I'd be careful not to mix too many scripts and config files in the `scripts` folder

File: src/main/java/prover/Prover.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+public class Prover {
+  public static void main(String[] args) throws IOException {
+    System.out.println("Distributed Prover (PoC implementation)");
+
+    var options = new Options();
+    options.addOption(new Option("h", "help", false, "Display this message"));
+    options.addOption(new Option("l", "local", false, "Run on local simulated cluster"));
+    options.addOption(new Option("t", "test", false, "Run trivial test to verify setup"));
+    options.addOption(new Option("p", "primary-size", false, "Size of primary input (1)"));
```
Comment (AntoineRondelet) [inline src/main/java/prover/Prover.java:37]: Shouldn't the `hasArg` flag be set to `true` here?

File: src/main/java/prover/Prover.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+      if (cmdLine.hasOption("test")) {
+        runTest(cmdLine.hasOption("local"));
+        return;
+      }
+
+      final int primaryInputSize = Integer.parseInt(cmdLine.getOptionValue("primary-size", "1"));
+      final String outputFile = cmdLine.getOptionValue("output", "proof.bin");
+
+      // Extract command lin arguments and call run.
```
Comment (AntoineRondelet) [inline src/main/java/prover/Prover.java:59]: command line

File: src/main/java/prover/Prover.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    final SparkSession spark = sessionBuilder.getOrCreate();
+
+    spark.sparkContext().conf().set("spark.files.overwrite", "true");
+    // spark.sparkContext().conf().set("spark.serializer",
+    // "org.apache.spark.serializer.KryoSerializer");
+    // spark.sparkContext().conf().registerKryoClasses(SparkUtils.zksparkClasses());
+    return new JavaSparkContext(spark.sparkContext());
+  }
+
+  static void runTest(boolean local) throws IOException {
```
Comment (AntoineRondelet) [inline src/main/java/prover/Prover.java:92]: I am not sure I get the purpose of this function. I can see that it is ran when the flag `--test` is set. What setup do you verify here?

Comment (dtebbs) [reply src/main/java/prover/Prover.java:94]: This is a simple test that the cluster is working and communicating correctly, which can be run standalone without any test data etc.

File: src/main/java/prover/Prover.java (hunk 3 above)
Comment (dtebbs) [reply src/main/java/prover/Prover.java]: Fixed typo in comment.

File: src/main/java/prover/Prover.java (hunk 2 above)
Comment (dtebbs) [reply src/main/java/prover/Prover.java]: Yes it should.  Fixed.  Thanks.

File: .github/workflows/main.yml (hunk 1 above)
Comment (dtebbs) [reply .github/workflows/main.yml]: Are you suggesting moving it somewhere else?  It's technically not a script by the strict definition, but pretty strongly tied to the `local-cluster-*` scripts, so it's not clear to me that we would gain anything by moving it.  I've changed these lines to not reference the file directly, so everything should be abstracted behind the `local-cluster-*` scripts.  But let me know if you prefer a different approach.  Thanks.

File: src/main/java/prover/Prover.java (hunk 4 above)
Comment (AntoineRondelet) [reply src/main/java/prover/Prover.java:94]: Got it, thanks

File: .github/workflows/main.yml (hunk 1 above)
Comment (AntoineRondelet) [reply .github/workflows/main.yml]: > pretty strongly tied to the local-cluster-* scripts, so it's not clear to me that we would gain anything by moving it. I've changed these lines to not reference the file directly, so everything should be abstracted behind the local-cluster-* scripts.

Makes sense, thanks. If the config file is only used by one script, that makes sense to keep it like that. My eye was simply caught when reading the CI config and seeing that some commands used config files in the `scripts` repo, which didn't seem appropriate.

**Review by AntoineRondelet — APPROVED** [review]
LGTM, thanks @dtebbs

**Pull request merged**